doc/manual/contract.md

```diff
@@ -0,0 +1,535 @@
+~~~nitish
+fun foo(a: Int, b: Int) is expect(a > 0), ensure(b < 0)
+~~~
+
+## Writing invariants
+
+Invariants are used to specify the characteristics of a class/interface which must always be true, except when executing a member function. In other words, invariants define conditions that must be respected by our instance. Any instance which does not respect one or more of those conditions will be considered as in an incoherent state, and will cause an "assertion failed" error at runtime.
+
+To define an invariant we use the class annotation `invariant`. Here is an example with our Stack. The class defines that it is impossible to be in a state where the size is less than 0.
+
+~~~nitish
+class Stack
+	invariant(size >= 0)
+end
+~~~
+
+Another example with a class that represents a date (day, month). To be in a consistent state, the date must be represented with a day between 1 and 31 and a month between 1 and 12.
+
+~~~
+class Date
+	invariant(day >= 1 and day <= 31)
+	invariant(month >= 1 and month <= 12)
+
+	var day: Int
+	var month: Int
+end
+~~~
+
+When a class defines an invariant all methods and constructors (inherited, redefined and introduced) will check this one. By default, the invariants are only checked on exit. See section `Verification policy` to activate them on in and out. For all constructors the invariant is only checked at the end of this, this is due to the fact that the object is not yet initialized. See the section `Execution order` for example.
+
+The invariant is not checked when the call is on `self`. The following example shows the execution of invariant:
+
+~~~
+class A
+	invariant(print_invariant)
+
+	fun print_invariant: Bool
+	do
+		print "In invariant verification"
+		return true
+	end
+
+	fun foo
+	do
+		print "In foo"
+		self.bar # The called property is a member of "self", the invariant will not be checked.
+	end
+
+	fun bar
+	do
+		print "In bar"
+	end
+end
+
+var a = new A
+a.foo # The called property is not a member of "self", the invariant will be checked.
+~~~
+
+This example gives the following output:
+~~~nitish
+In invariant verification
+In foo
+In bar
+In invariant verification
+~~~
+
+## Writing contracts
+
+Methods, as well as attribute setters, can be associated with two types of contracts; precondition (`expect`) that expresses requirements of the routine´s call and postcondition (`ensure`) that expresses what the routine guarantee after her execution. Together, they define the rules to be respected between the callee and the caller. If the caller fulfills the `expect` condition on entry then the caller can assume that the callee fulfill the `ensure` condition on exit.
+
+All contracts (`expect`, `ensure`) are evaluated in the same context as the called method, it's possible to refer to attributes, methods and parameters in the condition.
+
+When defining a contract (`expect`, `ensure`) on an attribute like bellow, only the automatically generated write property (setter) will use the contract.
```

Comment:
Éviter de faire des "like below" dans un texte, c'est moins convivial à lire. De plus, tu n'es pas obligé de spécifier que les setter sont automatiquement générer. Étant donné que les attributs sont une pair de getter/setter tu peux dire directement qu'il s'applique uniquement au setter. 

"Contract attributes are only applied to the write property (setter)"

Pas obligé de répéter que contract veut dire "expect" ou "ensure" une deuxième fois, tu le spécifies déjà plutôt.